File: leetcode/ag_05.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

char* longestPalindrome(char* s) {
    char ts[2005];
    memset(ts,0,sizeof(ts));
    ts[0] = '#';
    int cnt = 1;
    for(int i = 0; i < strlen(s); i++){
        ts[cnt++] = s[i];
        ts[cnt++] = '#';
    }
    int lents = strlen(ts);
    int max = 1;
    int pos = 0;
    for(int i = 0; i <lents; i++){
        int currentMax = 1;
        for(int j = 0 ; j < i ; j++){
            int l = i - j - 1;
            int r = i + j + 1;
            if(l < 0 || r >= lents){
                break;
            }
            if( ts[l] == ts[r]){
                currentMax+=2;
            }else {
                break;
            }
        }
        currentMax = (currentMax - 1) / 2;
        if (currentMax >= max){
            max = currentMax;
            pos = i;
        }
    }
    s = (char *)malloc(2005*sizeof(char));
    memset(s , 0 ,sizeof(s));
    cnt = 0;
    for(int i = pos - max + 1; i<pos + max ; i++){
        if(ts[i]!='#') s[cnt++] = ts[i];
    }
    s[cnt]='\0';
    return s;
}

int main()
{
    char str[1000];
    while(scanf("%s",str)!=EOF){
        char * ans = longestPalindrome(str);
        printf("%s\n",ans);
    }
    return 0;   
}

File: leetcode/ag_05.py
class Solution(object):

    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        ns = "#"+"#".join(list(s))+"#"
        max = 1
        pos = 0
        mlen = len(ns)
        for i in range(0,mlen):
            currentMax = 1;
            for j in range(0,i):
                l = i - j - 1
                r = i + 1 + j
                if l < 0 or r >= mlen:
                    break;
                if ns[l] == ns[r]:
                    currentMax+=2
                else:
                    break;
            currentMax = (currentMax - 1) /2
            if currentMax >= max:
                max = currentMax
                pos = i
        return "".join(ns[pos - max: pos + max + 1].split("#"))
    

    

def run():
    ns = Solution()
    st = "jglknendplocymmvwtoxvebkekzfdhykknufqdkntnqvgfbahsljkobhbxkvyictzkqjqydczuxjkgecdyhixdttxfqmgksrkyvopwprsgoszftuhawflzjyuyrujrxluhzjvbflxgcovilthvuihzttzithnsqbdxtafxrfrblulsakrahulwthhbjcslceewxfxtavljpimaqqlcbrdgtgjryjytgxljxtravwdlnrrauxplempnbfeusgtqzjtzshwieutxdytlrrqvyemlyzolhbkzhyfyttevqnfvmpqjngcnazmaagwihxrhmcibyfkccyrqwnzlzqeuenhwlzhbxqxerfifzncimwqsfatudjihtumrtjtggzleovihifxufvwqeimbxvzlxwcsknksogsbwwdlwulnetdysvsfkonggeedtshxqkgbhoscjgpiel"
    print ns.longestPalindrome(st)
